rascal-lsp/src/main/rascal/library/util/LanguageServer.rsc

```diff
@@ -270,8 +270,17 @@ data LanguageService
     | implementation(set[loc] (Focus _focus) implementationService)
     | codeAction    (list[CodeAction] (Focus _focus) codeActionService)
     | selectionRange(list[loc](Focus _focus) selectionRangeService)
+    | formatting    (list[TextEdit](Tree _input, loc range, FormattingOptions _opts) formattingService)
     ;
 
+data FormattingOptions(
```

Comment:
how do we know the default value? like if LSP does not send `trimTrailingWhitespace` (since it's an optional boolean) how do we know that means `true` and not `false`?

Reply:
I think the optional is only missing when either the server (which we control via our dependency version) or the client (which we do not control) are on an older version of the protocol.
I had a look at the protocol and LSP4J also provides some [default values](https://github.com/eclipse-lsp4j/lsp4j/blob/main/org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend#L4337-L4412). We could copy those - although I don't think `tabSize = 0` is a sensible default.

Reply:
The problem comes from typescript (and javascript) where `trimTrailingWhitespace?: boolean;` is a field that can become `undefined`, `false`, or `true`. But `undefined` is ["falsy"](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) so many places that would write `if (settings.trimTrailingWhitespace)` would mean that only if it's actually set to `true` does it have a value.

Also, more in this topic `tabSize` (and `insertSpaces`) is not optional in the [FormattingOptions](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#formattingOptions). So maybe it shouldn't be treated as a kw param?

Reply:
Not sure. When they added some of the optional properties for 3.15, they already had this: https://github.com/eclipse-lsp4j/lsp4j/commit/21c94d0b9536a9f8a31892c4134c7ab072999dc1#diff-7db90b832b0659f70a7dd341bb1ef0d9231ca767614dd73492c3e2d22d54f968

Maybe just a very defensive implementation?

Reply:
Discussed offline: we inspected LSP and VS Code behaviour and decided that the default values for optionals here should be `false`.